src/app/view/directives/dot-maxlength/dot-maxlength.directive.ts

```diff
@@ -7,29 +9,78 @@ import { Directive, ElementRef, HostListener, Input } from '@angular/core';
 @Directive({
     selector: '[dotMaxlength]'
 })
-export class DotMaxlengthDirective {
+export class DotMaxlengthDirective implements OnInit, OnDestroy {
     private _maxLength: number;
+    private events = ['paste', 'keypress'];
+    private destroy$: Subject<boolean> = new Subject<boolean>();
 
     constructor(private el: ElementRef) {}
 
+    ngOnInit() {
+        const eventStreams = this.events.map(ev => fromEvent(this.el.nativeElement, ev));
+        const allEvents$ = merge(...eventStreams);
+        allEvents$
+            .pipe(
+                takeUntil(this.destroy$),
+                tap((keyboardEvent: Event) => {
+                    if (!this.isValidAction(keyboardEvent)) {
+                        keyboardEvent.preventDefault();
+                    }
+                }),
+                delay(1),
+                filter(() => this.el.nativeElement.textContent.length > this._maxLength),
+                tap(() => this.reduceText())
+            )
+            .subscribe();
+    }
+
+    ngOnDestroy(): void {
+        this.destroy$.next(true);
+        this.destroy$.complete();
+    }
+
     @Input()
     set dotMaxlength(maxLength: number) {
         this._maxLength = maxLength || 255;
     }
 
+    private isValidAction(keyboardEvent: Event): boolean {
+        return (
+            this.el.nativeElement.textContent.length < this._maxLength ||
+            this.isAllowedKeyCode(keyboardEvent) ||
+            !!window.getSelection().toString()
+        );
+    }
+    private reduceText(): void {
+        this.el.nativeElement.textContent = this.el.nativeElement.textContent.slice(
+            0,
+            this._maxLength
+        );
+    }
+
     /**
-     * Listener fo the paste & keypress event. Will reduce the string to max Length.
+     * Check if a keycode is allowed when max limit is reached
+     * 8 : Backspace
+     * 37: LeftKey
+     * 38: UpKey
+     * 39: RightKey
+     * 40: DownKey
+     * ctrlKey for control key
+     * metakey for command key on mac keyboard
+     * @param {any} eventKeycode
+     * @returns boolean
      *
      * @memberof DotMaxlengthDirective
      */
-    @HostListener('paste', ['$event'])
-    @HostListener('keypress', ['$event'])
-    eventHandler() {
-        setTimeout(() => {
-            this.el.nativeElement.textContent = this.el.nativeElement.textContent.slice(
-                0,
-                this._maxLength
-            );
-        }, 0);
+    isAllowedKeyCode(event) {
+        return (
+            event.keyCode === 8 ||
```

Comment:
There is a property of the Keyboard event that is the string of the so you don't have to use a number and specify what it is in the doc.